// rele/dist/ql/parse.js
'use strict';

var _Object$defineProperty = require('babel-runtime/core-js/object/define-property')['default'];

_Object$defineProperty(exports, '__esModule', {
  value: true
});

exports['default'] = parse;

function parse(tokens, args) {
  var token = undefined;

  function nextToken() {
    token = tokens.shift();
  }

  function checkType(type) {
    return token && token[0] === type;
  }

  function assertType(type) {
    if (!checkType(type)) {
      throw new Error('Unexpected token "' + (token && token[0] || 'END') + '", expected ' + type);
    }
  }

  function readWord() {
    assertType('word');
    var word = token[1];
    nextToken();
    return word;
  }

  function readClassName() {
    var name = readWord();

    while (!checkType('{')) {
      if (checkType(':')) {
        name += readChar(':');
      } else {
        name += readWord();
      }
    }

    if (name[0].toUpperCase() !== name[0]) {
      throw new Error('Class name should start from capital letter "' + name + '"');
    }

    return name;
  }

  function readChar(char) {
    assertType(char);
    nextToken();
    return char;
  }

  function readField() {
    var name = readWord();
    return { type: 'field', name: name };
  }

  function readInclude() {
    var name = readWord();
    var cls = readAnonClass();
    return { type: 'include', name: name, 'class': cls };
  }

  function readPropety() {
    if (checkType('BREAK')) {
      var property = args.shift();
      if (property.type !== 'class') {
        throw new Error('Impossible injection "' + property + '"');
      }
      nextToken();
      return property;
    } else if (tokens[0] && tokens[0][0] === ':') {
      return readInclude();
    } else if (tokens[0] && tokens[0][0] === '{') {
      assertType('word');
      if (token[1][0].toUpperCase() === token[1][0]) {
        return readClass();
      } else {
        return readInclude();
      }
    } else if (tokens[0] && tokens[0][0] === '(') {
      return readCall();
    } else {
      return readField();
    }
  }

  function readProperties() {
    var properties = [];

    while (!checkType('}')) {
      properties.push(readPropety());

      if (checkType(',')) {
        readChar(',');
      }
    }

    return properties;
  }

  function readBlock() {
    readChar('{');
    var properties = readProperties();
    readChar('}');

    var fields = [];
    var includes = [];
    var classes = [];
    var calls = [];

    properties.forEach(function (property) {
      if (property.type === 'field') {
        fields.push(property.name);
      } else if (property.type === 'include') {
        includes.push(property);
      } else if (property.type === 'class') {
        classes.push(property);
      } else if (property.type === 'call') {
        calls.push(property);
      } else {
        throw new Error('You should never see this error');
      }
    });

    return { type: 'block', fields: fields, includes: includes, classes: classes, calls: calls };
  }

  function readNamedParam() {
    readChar('<');
    var name = readWord();
    readChar('>');

    return { type: 'param', name: name };
  }

  function readArg() {
    var key = readWord();
    var value = undefined;
    readChar(':');

    if (checkType('<')) {
      value = readNamedParam();
    } else {
      assertType('BREAK');
      value = args.shift();
      nextToken();
    }

    return { type: 'arg', key: key, value: value };
  }

  function readArgs() {
    var args = [];

    readChar('(');

    while (!checkType(')')) {
      args.push(readArg());
      if (checkType(',')) {
        readChar(',');
      }
    }

    readChar(')');

    return { type: 'args', args: args };
  }

  function readAnonClass() {
    var name = null;

    if (checkType(':')) {
      readChar(':');
      name = readClassName();
    }

    var block = readBlock();

    return { type: 'class', name: name, block: block };
  }

  function readCall() {
    var name = readWord();
    var args = readArgs();
    var cls = readAnonClass();
    return { type: 'call', name: name, args: args, 'class': cls };
  }

  function readClass() {
    var name = readClassName();
    var block = readBlock();
    return { type: 'class', name: name, block: block };
  }

  function readExpression() {
    if (tokens[0] && tokens[0][0] === '(') {
      return readCall();
    } else {
      return readClass();
    }
  }

  nextToken();
  return readExpression();
}

module.exports = exports['default'];
