// rele/dist/storage/jsonApi.js
'use strict';

var _Object$defineProperty = require('babel-runtime/core-js/object/define-property')['default'];

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

_Object$defineProperty(exports, '__esModule', {
  value: true
});

var _utilCollectJsonApiResources = require('../util/collectJsonApiResources');

var _utilCollectJsonApiResources2 = _interopRequireDefault(_utilCollectJsonApiResources);

function wrapper(obj) {
  return obj;
}

function merger(a, b) {
  var attributes = _Object$assign({}, a.attributes, b.attributes);
  var relationships = _Object$assign({}, a.relationships, b.relationships);

  return {
    type: a.type || b.type,
    id: a.id || b.id,
    attributes: attributes,
    relationships: relationships
  };
}

function getter(obj, key) {
  if (key === 'type' || key === 'id') {
    return obj[key];
  } else {
    return obj.attributes[key];
  }
}

function typeGetter(obj) {
  return obj.type;
}

function idGetter(obj) {
  return obj.id;
}

function relationshipGetter(obj, name) {
  return obj.relationships[name].data;
}

exports['default'] = {
  wrapper: wrapper,
  merger: merger,
  getter: getter,
  typeGetter: typeGetter,
  idGetter: idGetter,
  relationshipGetter: relationshipGetter,
  getResourcesFromResponse: _utilCollectJsonApiResources2['default']
};
module.exports = exports['default'];
